keps/sig-api-machinery/20190226-network-proxy.md

```diff
@@ -0,0 +1,618 @@
+It should also expose an administrative port speaking https. 
+This port is primarily for serving the liveness probe. 
+Without the liveness probe we can easily end up with a partially broken cluster 
+where the KAS cannot perform any of the requests requiring connections from the KAS 
+to the cluster. In addition this port could be used for serving things 
+such as pprof debug commands and monitoring data for the proxy. 
+
+### Direct Connection
+
+This connection type option is to allow for the use of the default dialer 
+if we want to bypass the connectivity proxy. 
+Among other things it allows for the connectivity service to be used 
+without actually using the connectivity proxy. 
+Once we roll out the feature this is a quick way to run the system 
+in a “legacy” mode or as a fallback. It also allows for very simple clusters 
+which do not need the concept of network segregation to be run 
+without the overhead (either in extra hop latency or configuration complexity) 
+of using the connectivity proxy.
+
+### Kubernetes API Server Outbound Requests
+
+The majority of communication which the KAS performs is based on incoming requests. 
+However there are a number of outgoing requests and those are the requests 
+which this document covers. The following is our understanding of those requests 
+and some details as to how they fit in this model.
+
+- **ETCD** It is possible to make etcd talk via the proxy. 
+The ectd client does take a transport. 
+(https://github.com/etcd-io/etcd/blob/master/client/client.go#L101) 
+so it is mostly a matter of adding configuration as to which proxy 
+an etcd client should use. 
+(https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/storage/storagebackend/config.go) 
+This would add an extra process hop to our main scaling axis. 
+Currently not considering it a requirement for this to be in place to proceed. 
+Suggest we scale test this. If no issues are found we proceed. 
+If a scale issue is found we may add an extra network 
+to allow for a configuration separate from ‘master’ should it be desired. 
+- **Pods/Exec**, **Pods/Proxy**, **Pods/Portforward**, **Pods/Attach**, **Pods/Log**
+We assume that all pod request (and pod subresource requests) are meant 
+for the cluster and will be routed based on the ‘cluster’ NetworkContext.
+- **Nodes/Proxy**
+We assume that all node request (and node subresource requests) 
+are meant for the cluster and will be routed based on the ‘cluster’ NetworkContext.
+- **Services/Proxy**
+We assume that all service request (and service subresource requests) 
+are meant for the cluster and will be routed based on the ‘cluster’ NetworkContext.
+- **Admission Webhooks**
+Admission webhooks can either be destined for a service or a URL. 
+If they are destined for a service then the service rules apply. 
+If they are destined for a URL then we will use the ‘master’ NetworkContext. 
```

Comment:
Can I not address a service by URL?

Reply:
I have never tried addressing a K8s service via the URL. If someone wants to send to a K8s service, they should use the service configuration. Given that the proxy used in these two examples will be different, they need to use the correct configuration.